### PR TITLE
feat: create user  api

### PR DESCRIPTION
**Pull request by mkhoi1998**
#### What's this PR do?

- [x] Add create user api
#### Screenshots (if appropriate)

image

### CODE REVIEW

File: pkg/view/employee.go [hunk 1]
```diff
@@ -54,6 +61,17 @@ type ProfileDataResponse struct {
 	Data ProfileData `json:"data"`
 }
 
+// CreateEmployeeRequest view for create new employee
+type CreateEmployeeRequest struct {
+	FullName      string            `json:"fullName" binding:"required"`
+	TeamEmail     string            `json:"teamEmail" binding:"required,email"`
+	PersonalEmail string            `json:"personalEmail" binding:"required,email"`
+	Role          string            `json:"role" binding:"required"`
```
Comment (namnhce) [inline pkg/view/employee.go:69]: This should be position

File: pkg/view/employee.go [hunk 2]
```diff
@@ -54,6 +61,17 @@ type ProfileDataResponse struct {
...
+	Salary        int               `json:"salary" binding:"required"`
+	Seniority     string            `json:"seniority" binding:"required"`
+	AccountRole   model.AccountRole `json:"accountRole" binding:"required"`
```
Comment (namnhce) [inline pkg/view/employee.go:72]:
```suggestion
	Role   model.Role `json:"role" binding:"required"`
```

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/employee/employee.go [hunk 3]
```diff
@@ -223,3 +225,66 @@ func (h *handler) GetProfile(c *gin.Context) {
...
+			ID: model.NewUUID(),
+		},
+		FullName:      req.FullName,
+		DisplayName:   strings.Join([]string{nameParts[len(nameParts)-1], nameParts[0]}, " "),
+		TeamEmail:     req.TeamEmail,
+		PersonalEmail: req.PersonalEmail,
+		WorkingStatus: model.WorkingStatusProbation,
+		JoinedDate:    &now,
+		AccountStatus: model.AccountStatusOnBoarding,
+		SeniorityID:   model.MustGetUUIDFromString(req.Seniority),
```
Comment (namnhce) [inline pkg/handler/employee/employee.go:270]: need to check seniority exists in db

File: pkg/handler/employee/employee.go [hunk 4]
```diff
@@ -223,3 +225,66 @@ func (h *handler) GetProfile(c *gin.Context) {
...
+		Roles: []model.Role{
```
Comment (namnhce) [inline pkg/handler/employee/employee.go:271]:
```suggestion
		Positions: []model.Position{
```

File: pkg/handler/employee/testdata/employee/list/without_working_status_and_pagination.json [hunk 5]
```diff
@@ -1,46 +1,76 @@
...
+				"createdAt": "0001-01-01T00:00:00Z",
+				"updatedAt": null,
+				"name": "",
+				"code": ""
+			},
+			"joinedDate": null,
+			"leftDate": null
+		}
+	]
+}
```
Comment (namnhce) [inline pkg/handler/employee/testdata/employee/list/without_working_status_and_pagination.json:76]: need empty line in the end of file

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/model/employees.go [hunk 6]
```diff
@@ -5,50 +5,62 @@ import "time"
 type WorkingStatus string
 
 const (
```
Comment (namnhce) [inline pkg/model/employees.go:7]: Need a function to check it's a valid value.
```
func (e WorkingStatus) IsValid() bool {
	switch e {
	case
		WorkingStatusPartTime,
		WorkingStatusLeft,
		WorkingStatusProbation,
		WorkingStatusFullTime:
		return true
	}
	return false
}
```

**Review by namnhce — CHANGES_REQUESTED** [review]

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/employee/employee.go [hunk 7]
```diff
@@ -290,3 +291,90 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+	})
+
+	// 1.2 prepare employee data
+	now := time.Now()
+	pos, err := h.store.Position.One(req.Position)
+	if err != nil {
+		l.Error(err, "error invalid position")
+		c.JSON(http.StatusBadRequest, view.CreateResponse[any](nil, nil, err, req))
+		return
+	}
```
Comment (namnhce) [inline pkg/handler/employee/employee.go:329]:
```suggestion
if err != nil {
		l.Error(err, "error invalid position")
		if gorm.IsByteArrayOrSlice(err) {
			c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, ErrPositionNotfound, req))
			return
		}
		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, err, req))
		return
	}
```

File: pkg/handler/employee/employee.go [hunk 8]
```diff
@@ -290,3 +291,90 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+	sen, err := h.store.Position.One(req.Seniority)
```
Comment (namnhce) [inline pkg/handler/employee/employee.go:330]:
```suggestion
	sen, err := h.store.Seniority.One(req.Seniority)
```

File: pkg/handler/employee/request.go [hunk 9]
```diff
@@ -17,3 +17,15 @@ type EditGeneralInfo struct {
...
+
+// CreateEmployee view for create new employee
+type CreateEmployee struct {
+	FullName      string     `json:"fullName" binding:"required,max=100"`
+	DisplayName   string     `json:"displayName" binding:"required"`
+	TeamEmail     string     `json:"teamEmail" binding:"required,email"`
+	PersonalEmail string     `json:"personalEmail" binding:"required,email"`
+	Position      model.UUID `json:"position" binding:"required"`
+	Salary        int        `json:"salary" binding:"required"`
+	Seniority     model.UUID `json:"seniority" binding:"required"`
```
Comment (namnhce) [inline pkg/handler/employee/request.go:29]:
```suggestion
	SeniorityID     model.UUID `json:"seniorityID" binding:"required"`
```

File: pkg/handler/employee/request.go [hunk 10]
```diff
@@ -17,3 +17,15 @@ type EditGeneralInfo struct {
...
 	NotionID      string `form:"notionID" json:"notionID"`
 }
+
+// CreateEmployee view for create new employee
+type CreateEmployee struct {
+	FullName      string     `json:"fullName" binding:"required,max=100"`
+	DisplayName   string     `json:"displayName" binding:"required"`
+	TeamEmail     string     `json:"teamEmail" binding:"required,email"`
+	PersonalEmail string     `json:"personalEmail" binding:"required,email"`
+	Position      model.UUID `json:"position" binding:"required"`
```
Comment (namnhce) [inline pkg/handler/employee/request.go:27]:
```suggestion
	PositionID      model.UUID `json:"positionID" binding:"required"`
```

File: pkg/model/employee_roles.go [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package model
+
+type EmployeeRole struct {
+	BaseModel
+
+	EmployeeID string
+	RoleID     string
+
+	Role Role `gorm:"foreignkey:role_id"`
```
Comment (namnhce) [inline pkg/model/employee_roles.go:9]: As I know, we don't need to use `gorm` tag here for FK declaration

File: pkg/store/employee/employee.go [hunk 12]
```diff
@@ -77,6 +82,12 @@ func (s *store) UpdateGeneralInfo(body EditGeneralInfo, id string) (*model.Emplo
 
 	// 1.3 save to DB
 	return employee, s.db.Table("employees").Where("id = ?", id).Updates(&employee).
-		Preload("Chapter").Preload("Seniority").Preload("LineManager").Preload("EmployeePositions").
-		Preload("EmployeePositions.Position").First(&employee).Error
+		Preload("Chapter").Preload("Seniority").Preload("LineManager").
```
Comment (namnhce) [inline pkg/store/employee/employee.go:85]:
```suggestion
		Preload("Chapter").
		Preload("Seniority").
		Preload("LineManager").
```

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/employee/employee.go [hunk 13]
```diff
@@ -290,3 +291,102 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
 	c.JSON(http.StatusOK, view.CreateResponse[any](view.ToEmployeeData(employee), nil, nil, nil))
 }
+
+// Create godoc
+// @Summary Create new employee
+// @Description Create new employee
+// @Tags Employee
+// @Accept  json
+// @Produce  json
+// @Success 200 {object} view.EmployeeData
```
Comment (datphamcode295) [inline pkg/handler/employee/employee.go:301]: add description for body

File: pkg/handler/employee/employee.go [hunk 14]
```diff
@@ -290,3 +291,102 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+		"params":  req,
+	})
+
+	// 1.2 prepare employee data
+	now := time.Now()
+	pos, err := h.store.Position.One(req.PositionID)
+	if err != nil {
+		l.Error(err, "error invalid position")
+		if err == gorm.ErrRecordNotFound {
+			c.JSON(http.StatusBadRequest, view.CreateResponse[any](nil, nil, ErrPositionNotfound, req))
```
Comment (datphamcode295) [inline pkg/handler/employee/employee.go:328]:
```suggestion
			c.JSON(http.StatusNotFound, view.CreateResponse[any](nil, nil, ErrPositionNotfound, req))
```

File: pkg/handler/employee/employee.go [hunk 15]
```diff
@@ -290,3 +291,102 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+			return
+		}
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, err, req))
+		return
+	}
+	sen, err := h.store.Seniority.One(req.SeniorityID)
+	if err != nil {
+		l.Error(err, "error invalid seniority")
+		if err == gorm.ErrRecordNotFound {
+			c.JSON(http.StatusBadRequest, view.CreateResponse[any](nil, nil, ErrSeniorityNotfound, req))
```
Comment (datphamcode295) [inline pkg/handler/employee/employee.go:338]:
```suggestion
			c.JSON(http.StatusNotFound, view.CreateResponse[any](nil, nil, ErrSeniorityNotfound, req))
```

File: pkg/handler/employee/employee.go [hunk 16]
```diff
@@ -290,3 +291,102 @@ func (h *handler) UpdateGeneralInfo(c *gin.Context) {
...
+	employee := &model.Employee{
+		BaseModel: model.BaseModel{
+			ID: model.NewUUID(),
+		},
+		FullName:      req.FullName,
+		DisplayName:   req.DisplayName,
+		TeamEmail:     req.TeamEmail,
+		PersonalEmail: req.PersonalEmail,
+		WorkingStatus: model.WorkingStatusProbation,
+		JoinedDate:    &now,
```
Comment (datphamcode295) [inline pkg/handler/employee/employee.go:363]: I'm not sure JoneDate is the create record time or the time employee become Dwarves's member ? cc anh @namnhce

File: pkg/handler/employee/employee_test.go [hunk 17]
```diff
@@ -251,3 +254,103 @@ func Test_UpdateGeneralInfo(t *testing.T) {
...
+				RoleID:        model.MustGetUUIDFromString("d796884d-a8c4-4525-81e7-54a3b6099eac"),
+			},
+			id: "2655832e-f009-4b73-a535-64c3a22e558f",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			byteReq, err := json.Marshal(tt.body)
+			fmt.Println(err)
```
Comment (namnhce) [inline pkg/handler/employee/employee_test.go:329]: remove this one

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/employee/employee.go (hunk 16 above)
Comment (namnhce) [reply pkg/handler/employee/employee.go:363]: for now we use current time to determine joined date.

**Review by namnhce — APPROVED** [review]

**Pull request merged**